### PR TITLE
[TorchComms + DeviceMesh] Use backend Wrapper on the c10d level for devicemesh + TorchComms

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170132

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -234,6 +234,7 @@ def __init__(
             self._rank_map = _rank_map
             self._mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
             self._root_mesh = _root_mesh
+            self._use_torchcomm = True if os.environ.get("TORCH_USE_COMMS_BACKEND") else False
```
Comment (d4l3k) [inline torch/distributed/device_mesh.py]: if this is a global should we just reference it statically rather than having to pass it through everywhere?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -2034,6 +2038,15 @@ def _new_process_group_helper(
                     backend_class.size(),
                 )
                 pg._set_default_backend(backend_type)
+        elif _use_torchcomm:
+            torchcomm_backend = os.environ["TORCHCOMM_BACKEND"]
```
Comment (d4l3k) [inline torch/distributed/distributed_c10d.py]: This seems like it'll cause problems if users want to use gloo and nccl.

Can we not use backend_str as usual? Could also name the backends like `torchcomms:nccl` or something to trigger via this path.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -5256,13 +5270,19 @@ def split_group(
...
-        group_name=group_name,
-        group_desc=group_desc,
-    )
+    if _use_torchcomm:
+        parent_comm = parent_pg.get_comm()
+        split_comm = parent_comm.split(my_group, group_name)
+        from torchcomms._comms import _BackendWrapper, new_comm
+        split_pg = _BackendWrapper(split_comm)
+    else:
+        split_pg = parent_pg.split_group(
```
Comment (d4l3k) [inline torch/distributed/distributed_c10d.py]: we should just implement _BackendWrapper.split_group ?

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/distributed_c10d.py]: but torchComms does not support that right? users need to always create different objects. But I see what you mean. The signature of new_comm needs a device.

File: torch/distributed/device_mesh.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/device_mesh.py]: you mean like a module level flag of device_mesh?

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/distributed_c10d.py]: OK I see what you mean.. One thing we need to solve is how to map nccl to ncclx?

Comment (fduwjj) [reply torch/distributed/distributed_c10d.py]: We might also need to update map `backend_type_map`?

**Review by d4l3k — APPROVED** [review]
I think this is fine for now -- I still don't love the _use_torchcomms being passed around and would prefer subclass approach but we can change that later

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -2034,6 +2038,14 @@ def _new_process_group_helper(
                     backend_class.size(),
                 )
                 pg._set_default_backend(backend_type)
+        elif _use_torchcomm:
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py]: i'm not sure what option we have besides adding a torchcomm path here.  however, we need to align all the kwargs of _new_process_group_helper to support torchcomms equivalents, and it gets messy in certain cases, like backend_options having different formats.  Any thoughts about how to clean this up?

Comment (fduwjj) [reply torch/distributed/distributed_c10d.py]: great point, we can add a option conversion here. Let me update it.. Was caught up by other things yesterday.

File: test/distributed/test_device_mesh.py [hunk 5]
```diff
@@ -1458,6 +1458,44 @@ def test_scatter_nd(self):
             mesh_scatter(received_tensor, scattered_tensors, mesh, mesh_dim=dim)
             self.assertEqual(received_tensor, torch.ones(3, 3) * self.rank)
 
+    @with_comms
+    def test_device_mesh_w_torchcomms(self) -> None:
+        DeviceMesh._use_torchcomm = True
+        os.environ["TORCH_BACKEND"] = "ncclx"
```
Comment (wconstab) [inline test/distributed/test_device_mesh.py]: wondering why we need ncclx passed to an ENV when we are also passing it to backend_override?

File: torch/distributed/distributed_c10d.py [hunk 6]
```diff
@@ -141,6 +141,14 @@
 _UCC_AVAILABLE = True
 _XCCL_AVAILABLE = True
 
+try:
+    # pyrefly: ignore [missing-import]
+    from torchcomms._comms import _BackendWrapper, new_comm
+
+    _TORCHCOMM_AVAILABLE = True
+except ImportError:
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:149]: using try/except importerror is considered an antipattern.  it can mask other errors.  instead, i think you can use `if importlib.util.find_spec("torchcomms")`

File: torch/distributed/distributed_c10d.py [hunk 7]
```diff
@@ -2038,8 +2046,11 @@ def _new_process_group_helper(
...
                 )
                 pg._set_default_backend(backend_type)
-        elif _use_torchcomm: 
-            from torchcomms._comms import _BackendWrapper, new_comm
+        elif _use_torchcomm:
+            if not _TORCHCOMM_AVAILABLE:
+                raise RuntimeError(
+                    "torchcomms is not available. Please install torchcomms to use this feature."
+                )
             device = torch.device(os.environ.get("TORCHCOMM_DEVICE", "cuda"))
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py]: why does torchcomm get its device from an env instead of the devicemesh API?

File: test/distributed/test_device_mesh.py (hunk 5 above)
Comment (fduwjj) [reply test/distributed/test_device_mesh.py]: this is for init_process_group.

File: torch/distributed/distributed_c10d.py (hunk 6 above)
Comment (fduwjj) [reply torch/distributed/distributed_c10d.py:149]: sure, I think I can do that.

File: torch/distributed/distributed_c10d.py (hunk 7 above)
Comment (fduwjj) [reply torch/distributed/distributed_c10d.py]: oh right we can use the device_type from DeviceMesh, good point.

File: test/distributed/test_device_mesh.py (hunk 5 above)
Comment (wconstab) [reply test/distributed/test_device_mesh.py]: Hmm, I'm still confused, why can't the backend override be used for init?

Comment (kapilsh) [reply test/distributed/test_device_mesh.py]: I was confused by this too -- not sure why we need an env variable

File: torch/distributed/distributed_c10d.py [hunk 8]
```diff
@@ -5290,6 +5344,9 @@ def split_group(
         for group_rank, global_rank in enumerate(global_ranks_in_my_group)
     }
 
+    if _use_torchcomm:
+        # pyrefly: ignore [missing-attribute]
+        _world.add_comm(split_backend_class.get_comm())
```
Comment (kapilsh) [inline torch/distributed/distributed_c10d.py:5349]: this seems suspicious.
